# knitr document van Steensel lab

# Gene repression in LADs
## Christ Leemans, 03-11-2016 - to date

## Introduction
Generally speaking, genes inside lamina associated domains are not or very lowly expressed. These genes can either be actively repressed by their DNA context (e.g. heterochromatin, lamina association), or simply be inactive (because essential factors for expression are missing?). Yet another group of genes seem to evade gene repression in the context of lamina associated domains. Previously we devided promoters in these three groups. Now I would like to see if there is any particular organisation of these promoters within LADs, are these groups clustering together, or are they interspersed in a way such that the inactive promoters can help repress the repressed promoters.

## Description of Data.

gencode.sure.160617.rda: 
    file from Joris, received 17 June 2016. Promoter positions in this file are from Gencode. Contains SuRE, gro-cap and cage expression data as well as the number of tissues in which each promoter is expressed.


## libraries, paths and data prep


```{r, fig.width=10, fig.height=10}
library(reshape2)
library(rtracklayer)
library(ggplot2)
library(DESeq2)
library(gridExtra)
library(plyr)
library(scales)
library(Matrix)
library(grid)
library(glmnet)
library(rtfbs)
library(BSgenome)
library(BSgenome.Hsapiens.UCSC.hg19)
library(doMC)
library(ggdendro)
registerDoMC(cores=10)
load('../raw_data/biomart.rdata')

## There was one promoter that was wrongly annotated
bm_p[bm_p$ensembl_transcript_id=='ENST00000357491','ensembl_gene_id' ] = 'ENSG00000196350'
```



### SuRE data
Previously, Bas defined three different groups of promoters in LADs. The "escaper", "repressed" and "inactive" group. The "escaper" group contains promoters that are expressed in their endogenous LAD environment. The "repressed" group is inactive in the endogenous setting, but they do show autonomous promoter activity in SuRE data. The 'inactive' group shows neither endogenous expression nor autonomous promoter activity in SuRE data.

```{r, fig.width=10, fig.height=8}
load("../raw_data/gencode.sure.160617.rda")
Prom<-gencode.sure.160617; rm(gencode.sure.160617) #simpler name
#first re-calculate pseudocounts without jitter
P<-Prom[,c(1:8,23,20,26,28, 27)] #SuRE, GRO-cap, CAGE and LAD columns only
names(P)[9:13]<-c("SuRE", "GROcap", "CAGE", "LAD", 'tissues_expressed')


## for promoters and gene expression let's convert promoter transcript id's to gene id's
P$ensembl_transcript_id = do.call(rbind, strsplit(P$name, split='[.]'))[,1]

nrow(P) #orriginal number of rows
bm_match = match(P$ensembl_transcript_id, bm_p$ensembl_transcript_id)
P<-merge(P, bm_p, by="ensembl_transcript_id", all.x=TRUE)
nrow(P) #some double rows were introduced

P = P[match(Prom$name, P$name), ]

length(unique(P$ensembl_gene_id)) #number of unique genes

table(P[,c('strand.x','strand.y')]) #almost all strand listings are consistent

P<-P[, colnames(P)!='strand.y']
colnames(P)[colnames(P)=='strand.x'] = "strand"



## to be used by CGtools as the complete set of TSS's
peaks = data.frame(seqname=P$chr,
                   start=P$tss,
                   end=P$tss,
                   strand=P$strand)


Pseud<-min(P$SuRE[P$SuRE>0], na.rm=TRUE)/2
P$SuRE<-P$SuRE+Pseud
P$SuRE<-log10(P$SuRE)
PseudGro<-min(P$GROcap[P$GROcap>0], na.rm=TRUE)/2
P$GROcap<-P$GROcap+PseudGro
P$GROcap<-log10(P$GROcap)
PseudCage<-min(P$CAGE[P$CAGE>0], na.rm=TRUE)/2
P$CAGE<-P$CAGE+PseudCage
P$CAGE<-log10(P$CAGE)

#then calculate running mean for iLAD promoters:
P<-P[order(P$SuRE,sample(c(1:nrow(P)))),] #sort by SuRE and then random for ties
n<-60 #number of windows
w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
s<-round(seq(from=w/2+0.0001, to=nrow(P)-w/2, length.out=n))
RM<-data.frame(SuRE.low=rep(NA,n), SuRE.mean=rep(NA,n), SuRE.hi=rep(NA,n), GROcap.lad=rep(NA,n), GROcap.ilad=rep(NA,n))
RM$SuRE.low=P$SuRE[s-floor(w/2)]
for(i in 1:n){RM$SuRE.mean[i]=mean(P$SuRE[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
RM$SuRE.hi=P$SuRE[s+floor(w/2)]
for(i in 1:n)
  {t<-P[(s[i]-floor(w/2)):(s[i]+floor(w/2)),]
   RM$GROcap.lad[i]<-mean(t$GROcap[t$LAD==1], na.rm=TRUE)
   RM$GROcap.ilad[i]<-mean(t$GROcap[t$LAD==0], na.rm=TRUE)
  }


#add first datapoint (SuRE equals pseudocount)
RM1<-RM[0,] #empty df
RM1[1,]<-c(rep(log10(Pseud),3), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==1]), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==0]))
RM<-rbind(RM1, RM)
rm(RM1)

#finally calculate LRS for all promoters:
P$LRS<- P$GROcap - approx(x=RM$SuRE.mean, y=RM$GROcap.ilad, xout=P$SuRE, rule=2)$y
#so the more negative the score, the more 'repressed' is the promoter by its chromatin/LAD context


#(arbitrary) cutoffs to define three groups of promoters:
INACT<- P$SuRE< -0.3 & P$LAD==1 & P$GROcap< -2 #inactive
NREP<- P$SuRE> 0 & P$LRS> -0.5 & P$LAD==1 & P$GROcap> -2 #not repressed
REP<- P$SuRE> 0.3 & P$LRS< -1 & P$LAD==1  & P$GROcap< -2 #repressed
Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
names(Pcnts)<-c("repressed", "escaper", "inactive")
BND <- P$LAD==1 & !INACT & !NREP & !REP

#add class annotation column to P:
P$class<-NA
P$class[P$LAD==0]<-"iLAD"
P$class[INACT]<-"inactive"
P$class[NREP]<-"escaper"
P$class[REP]<-"repressed"
P$class[BND] <- "boundary"
P$class = factor(P$class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary'))

COLi<-"#00BBFF11" #dot color for iLAD promoters
COL_lad<-c("#FF0000", "#0077FF")
names(COL_lad)<-c('LAD', 'iLAD')

#color vector for plotting:
COL_class<-c("#A020F0", "#FFA500", "#006400", "#7e7e7e", "#0077FF")
names(COL_class)<-c("repressed", "escaper", "inactive", 'boundary', 'iLAD')

COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "escaper", "inactive")

class_names = paste0(levels(P$class), '; n=',table(P$class))
names(class_names) = levels(P$class)
P$class_n_prom = P$class
levels(P$class_n_prom) = class_names
COL_class_n_prom = COL_class[names(class_names)]
names(COL_class_n_prom) = class_names

lad_names = c(LAD=paste0('LAD; n=', table(P$LAD)['1']),
              iLAD=paste0('LAD; n=', table(P$LAD)['0']))
P$lad_n_prom = factor(ifelse(P$LAD==1, lad_names['LAD'], lad_names['iLAD']))
COL_lad_n = COL_lad
names(COL_lad_n) = lad_names


RM_melt = melt(RM, measure.vars=c('GROcap.ilad', 'GROcap.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='GROcap.lad', lad_names['LAD'], lad_names['iLAD'])
ggplot(P, aes(x=SuRE, y=GROcap, color=lad_n_prom)) +
    geom_point(data=P[P$LAD==0, ], size=0.5, alpha=0.05) + 
    geom_point(data=P[P$LAD==1, ], size=0.5, alpha=0.2) + 
    theme_bw() +
    geom_line(data=RM_melt, aes(x=SuRE.mean, y=value, color=variable), size=1) +
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_n)
p_classes = P[which(P$class %in% c('inactive', 'escaper', 'repressed')),]
ggplot(P, aes(x=SuRE, y=GROcap)) + 
    geom_point(size=0.1,color=COLi) + 
    geom_point(data=p_classes, aes(color=class_n_prom), size=0.6) + 
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme_bw() +
    geom_line(data=RM, aes(x=SuRE.mean, y=GROcap.ilad), color=COL_lad['iLAD']) +
    theme(legend.title=element_blank()) +
    scale_colour_manual(values=COL_class_n_prom) 

```

**conclusion:**

We now have a definition of 3 different groups. We can now look at what is different between these groups.


## create GRanges

```{r}
P_per_gene = ddply(P, .(ensembl_gene_id), function(x){
        return(x[sample(1:nrow(x), 1), ])
    })

count_kmer_in_lad <- function(o_table, k=2){
    lad_class_order = list()
    lad_strand_order = list()
    unique_lad = unique(o_table$queryHits)
    lad_n = length(unique_lad)
    for (i in 1:lad_n){
        class_vec = o_table$class[which(o_table$queryHits==unique_lad[i])]
        strand_vec = o_table$strand[which(o_table$queryHits==unique_lad[i])]
        lad_class_order[[i]] = cbind(class=class_vec, strand=strand_vec)
    }
    k_mer_list = lapply(lad_class_order, function(x){
            if (nrow(x) >= k){
                l = nrow(x)-k+1
                comb_vec = vector(length=l)
                j = 1
                while (j <= l){
                    comb = paste(x[j:(j+k-1),'class'],collapse='_')
                    if (k==2){
                        s1 = x[j,'strand']
                        s2 = x[j+1, 'strand']
                        # if (s1 == s2){
                        #     comb = paste(comb, 'same_strand',sep='_')
                        # } else if (s1=='-'){
                        #     comb = paste(comb, 'outwards',sep='_')
                        # } else {
                        #     comb = paste(comb, 'inwards',sep='_')
                        # }
                    }
                    comb_vec[j] = comb
                    j = j + 1
                }
                return(comb_vec)
            }
        })
    return(table(unlist(k_mer_list)))
}
#genes as Granges object
tss_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P_per_gene$chr,
                                              start=P_per_gene$tss,
                                              end=P_per_gene$tss,
                                              strand=P_per_gene$strand,
                                              class=P_per_gene$class),
                                              keep.extra.columns=TRUE)

lad_gr = import.bed('../../../data/tracks/hg19/cl20161019_LAD_continuous_2state_K562.bed')

tss_gr = tss_gr[order(seqnames(tss_gr), start(tss_gr))]


o = findOverlaps(lad_gr[lad_gr$name=='LAD'], tss_gr)

o_table = data.frame(o)
o_table$class = as.character(tss_gr[o_table$subjectHits]$class)
o_table$strand = as.vector(strand(tss_gr))[o_table$subjectHits]

real_kmer = count_kmer_in_lad(o_table)

sim_list_kmer = lapply(1:1000, function(x, o_table){
        o_table$class = sample(o_table$class)
        count_kmer_in_lad(o_table)
    }, o_table)

col_vec = rainbow(length(real_kmer))

option_vec = c()
for (class1 in c('escaper', 'repressed', 'inactive', 'boundary')){
    for (class2 in c('escaper', 'repressed', 'inactive', 'boundary')){
        # for (mode in c('same_strand', 'inwards', 'outwards')){
            option_vec = c(option_vec, paste(class1, class2, sep='_'))
        # }
    }
}

sim_table_kmer = matrix(0, nrow=length(option_vec), ncol=1000,
                        dimnames=list(option_vec))
for (i in 1:1000){
    sim_table_kmer[names(sim_list_kmer[[i]]),i] = sim_list_kmer[[i]]
}

sim_melt = melt(sim_table_kmer)[,-2]
real_table = data.frame(real_kmer)
colnames(sim_melt) = colnames(real_table) = c('dimer', 'count')
ggplot(sim_melt, aes(x=count, color=dimer)) +
    geom_density() +
    geom_vline(data=real_table, aes(xintercept=count)) +
    facet_wrap(~dimer) +
    ggtitle('dimer counts of subsequent promoters in each LAD\nvertical lines for real counts\nwith density plot of 1000 randomly shuffled classifications')


real_kmer = count_kmer_in_lad(o_table, 3)

sim_list_kmer = lapply(1:1000, function(x, o_table){
        o_table$class = sample(o_table$class)
        count_kmer_in_lad(o_table, 3)
    }, o_table)

col_vec = rainbow(length(real_kmer))

sim_table_kmer = do.call(rbind, sim_list_kmer)


sim_melt = melt(sim_table_kmer)[,-1]
real_table = data.frame(real_kmer)
colnames(sim_melt) = colnames(real_table) = c('trimer', 'count')
ggplot(sim_melt, aes(x=count, color=trimer)) +
    geom_density() +
    geom_vline(data=real_table, aes(xintercept=count)) +
    facet_wrap(~trimer) +
    ggtitle('trimer counts of subsequent promoters in each LAD\nvertical lines for real counts\nwith density plot of 1000 randomly shuffled classifications')



max_ratio_in_lad <- function(o_table, k=2){
    lad_class_order = list()
    unique_lad = unique(o_table$queryHits)
    lad_n = length(unique_lad)
    for (i in 1:lad_n){
        class_vec = o_table$class[which(o_table$queryHits==unique_lad[i])]
        lad_class_order[[i]] = class_vec
    }

    max_ratio_list = lapply(lad_class_order, function(x){
            if (length(x) >= k){
                return(max(table(x) / length(x)))
            } else {
                return(NaN)
            }
        })
    max_ratio_vec = unlist(max_ratio_list)
    return(max_ratio_vec[!is.na(max_ratio_vec)])
}



real_max_ratio = max_ratio_in_lad(o_table, 4)

sim_list_max_ratio = lapply(1:1000, function(x, o_table){
        o_table$class = sample(o_table$class)
        max_ratio_in_lad(o_table, 4)
    }, o_table)

mean_vec = unlist(lapply(sim_list_max_ratio, mean))

plot(density(mean_vec), main='mean ratio of highest occuring class in lad\ndensity plot of 1000 simulations\nand vertical line at actual mean ratio')
abline(v=mean(real_max_ratio))

single_vec = unlist(lapply(sim_list_max_ratio, function(x){length(which(x==1))}))
real_single = length(which(real_max_ratio==1))
plot(density(single_vec), main='count of LADs with only a single class of promoter\ndensity plot of 1000 simulations\nand vertical line at actual count')
abline(v=real_single)


rel_pos_in_lad <- function(o_table, k=2){
    lad_class_order = list()
    unique_lad = unique(o_table$queryHits)
    lad_n = length(unique_lad)
    for (i in 1:lad_n){
        class_vec = o_table$class[which(o_table$queryHits==unique_lad[i])]
        lad_class_order[[i]] = class_vec
    }

    rel_pos_list = lapply(lad_class_order, function(x){
            if (length(x) >= k){
                i_vec = 1:length(x)
                rel_pos_vec = i_vec - 0.5 - length(i_vec)/2
                r = data.frame(class=x, rel_pos=abs(rel_pos_vec/max(rel_pos_vec)))
                return(r)
            }
        })
    rel_pos_table = do.call(rbind.data.frame, rel_pos_list)
    return(rel_pos_table)
}


real = rel_pos_in_lad(o_table, 4)

ggplot(real, aes(x=class, y=rel_pos, color=class)) +
    geom_violin(alpha=0.5) +
    geom_point(position=position_jitter(width=0.5), alpha=0.3, size=0.5) +
    scale_color_manual(values=COL_class)

```

**conclusion:**
Maybe not the best visualizations, but it is clear that promoters of the same class tend to cluster together inside LADs and there is no specific order from lad boundary towards the center (e.g. all escapers near edge of the LAD and inactive genes in the centre). There is no higher occurence of LADs with only a single type of promoter in it.

Probably the higher occurence of "dimers" of promoters with the same class is due to some promoters that lay in close proximity and initiate trascription in opposing directions.


## differences between constitutive and facultative LADs
I was wondering how different classes were organised across consitutive and facultative LADs.

I looked at a lad state definition previously generated by myself. I took carolyn's data (see DokuWiki) to generate continuous tracks with ciLAD,fiLAD,fLAD and cLAD regions. I excluded data from both twins in the down syndrome fibroblasts and data obtained from Guelen et al. since Guelen et al. was created using a different microarray platform and the down syndrome fibroblast data contained some weird stuff.

```{r}

count_kmer_in_lad <- function(o_table, k=2, use_ori=F){
    lad_class_order = list()
    lad_strand_order = list()
    unique_lad = unique(o_table$queryHits)
    lad_n = length(unique_lad)
    for (i in 1:lad_n){
        class_vec = o_table$class[which(o_table$queryHits==unique_lad[i])]
        strand_vec = o_table$strand[which(o_table$queryHits==unique_lad[i])]
        start_vec = o_table$start[which(o_table$queryHits==unique_lad[i])]
        lad_class_order[[i]] = data.frame(class=class_vec, strand=strand_vec, start=start_vec)
    }
    kmer_list = lapply(lad_class_order, function(x){
            if (nrow(x) >= k){
                l = nrow(x)-k+1
                comb_vec = vector(length=l)
                dist_vec = vector(length=l)
                j = 1
                while (j <= l){
                    comb = x[j:(j+k-1),'class']
                    start_vec = x[j:(j+k-1),'start']
                    if (order(comb[c(1,k)])[1]==2){
                        comb = comb[k:1]
                    }
                    comb_str = paste(comb,collapse='_')
                    if (k==2 & use_ori){
                        s1 = x[j,'strand']
                        s2 = x[j+1, 'strand']
                        if (s1 == s2){
                            comb_str = paste(comb_str, 'same_strand',sep='_')
                        } else if (s1=='-'){
                            comb_str = paste(comb_str, 'outwards',sep='_')
                        } else {
                            comb_str = paste(comb_str, 'inwards',sep='_')
                        }
                    }
                    comb_vec[j] = comb_str
                    dist_vec[j] = start_vec[k] - start_vec[1]
                    j = j + 1
                }
                return(data.frame(combination=comb_vec, dist_vec))
            }
        })
    kmer_table = do.call(rbind.data.frame, kmer_list)
    result = ddply(kmer_table, .(combination), function(x){
            c(count=nrow(x), mean_dist=round(sum(x$dist_vec)/nrow(x)))
    })
    result$combination = as.character(result$combination)
    return(result)
}
#genes as Granges object
tss_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P_per_gene$chr,
                                              start=P_per_gene$tss,
                                              end=P_per_gene$tss,
                                              strand=P_per_gene$strand,
                                              class=P_per_gene$class,
                                              name=P_per_gene$name),
                                              keep.extra.columns=TRUE)

lad_gr = import.bed('../../../data/tracks/hg19/cl20161019_LAD_continuous_2state_K562.bed')

tss_gr = tss_gr[order(seqnames(tss_gr), start(tss_gr))]


o = findOverlaps(lad_gr[lad_gr$name=='LAD'], tss_gr)

o_table = data.frame(o)
o_table$class = as.character(tss_gr[o_table$subjectHits]$class)
o_table$strand = as.vector(strand(tss_gr))[o_table$subjectHits]
o_table$start = as.vector(start(tss_gr))[o_table$subjectHits]
o_table$name = as.character(tss_gr[o_table$subjectHits]$name)

real_kmer = count_kmer_in_lad(o_table)

sim_list_kmer = lapply(1:1000, function(x, o_table){
        o_table$class = sample(o_table$class)
        count_kmer_in_lad(o_table)
    }, o_table)

col_vec = rainbow(length(real_kmer))

option_vec = c()
class_vec = c('escaper', 'repressed', 'inactive', 'boundary')
class_vec = class_vec[order(class_vec)]
for (i in 1:length(class_vec)){
    for (j in i:length(class_vec)){
        # for (mode in c('same_strand', 'inwards', 'outwards')){
            option_vec = c(option_vec, paste(class_vec[i], class_vec[j], sep='_'))
        # }
    }
}

sim_table_kmer = matrix(0, nrow=length(option_vec), ncol=1000,
                        dimnames=list(option_vec))
for (i in 1:1000){
    sim_table_kmer[sim_list_kmer[[i]]$combination,i] = sim_list_kmer[[i]]$count
}

sim_melt = melt(sim_table_kmer)[,-2]
colnames(sim_melt) = colnames(real_table) = c('combination', 'count')
ggplot(sim_melt, aes(x=count, color=combination)) +
    geom_density() +
    geom_vline(data=real_kmer, aes(xintercept=count)) +
    facet_wrap(~combination) +
    ggtitle('dimer counts of subsequent promoters in each LAD\nvertical lines for real counts\nwith density plot of 1000 randomly shuffled classifications')

dist_table_kmer = matrix(0, nrow=length(option_vec), ncol=1000,
                        dimnames=list(option_vec))
for (i in 1:1000){
    dist_table_kmer[sim_list_kmer[[i]]$combination,i] = sim_list_kmer[[i]]$mean_dist
}

dist_melt = melt(dist_table_kmer)[,-2]
colnames(dist_melt) = colnames(real_table) = c('combination', 'mean_dist')
ggplot(dist_melt, aes(x=mean_dist, color=combination)) +
    geom_density() +
    geom_vline(data=real_kmer, aes(xintercept=mean_dist)) +
    facet_wrap(~combination) +
    ggtitle('dimer counts of subsequent promoters in each LAD\nvertical lines for real counts\nwith density plot of 1000 randomly shuffled classifications')


```


```{r}
lad_4state = import.bed('../../../data/tracks/hg19/cl20161017_LAD_continuous_4state_K562.bed')
tss_gr_4state <-makeGRangesFromDataFrame(data.frame(seqnames=P$chr,
                                              start=P$tss,
                                              end=P$tss,
                                              strand=P$strand,
                                              class=P$class),
                                              keep.extra.columns=TRUE)
o_4state = findOverlaps(tss_gr_4state, lad_4state)
P$lad_4state = NA
P$lad_4state[queryHits(o_4state)] = lad_4state$name[subjectHits(o_4state)]
plot_list = list()
for (lad_state in c('ciLAD', 'fiLAD', 'fLAD', 'cLAD')){
    plot_list[[lad_state]] = ggplot(P[P$lad_4state==lad_state,], aes(x=class, fill=class)) +
                                geom_bar() +
                                ggtitle(lad_state) +
                                geom_bar(aes(y = (..count..)/sum(..count..))) +
                                scale_y_continuous(labels=percent, limits=c(0,1)) +
                                scale_fill_manual(values=COL_class)
} 


```

**conclusion:**

There is a higher ratio of inactive genes in constitutive LADs. Think these genes are verry tissue specific and will only detach from the LAD in some specific cell types.



```{r}
kmer_in_lad <- function(o_table, k=2){
    lad_class_order = list()
    lad_strand_order = list()
    unique_lad = unique(o_table$queryHits)
    lad_n = length(unique_lad)
    for (i in 1:lad_n){
        class_vec = o_table$class[which(o_table$queryHits==unique_lad[i])]
        strand_vec = o_table$strand[which(o_table$queryHits==unique_lad[i])]
        start_vec = o_table$start[which(o_table$queryHits==unique_lad[i])]
        name_vec = o_table$name[which(o_table$queryHits==unique_lad[i])]
        lad_class_order[[i]] = data.frame(class=class_vec, strand=strand_vec, start=start_vec,
                                          name=name_vec)
    }
    kmer_list = lapply(lad_class_order, function(x){
            if (nrow(x) >= k){
                l = nrow(x)-k+1
                comb_vec = vector(length=l)
                dist_vec = vector(length=l)
                comb_frame = data.frame(class1=rep('',l), class2=rep('',l), stringsAsFactors=F)
                name_frame = data.frame(name1=rep('',l), name2=rep('',l), stringsAsFactors=F)
                if (k==2){
                    comb_frame$ori = rep('',l)
                }
                j = 1
                while (j <= l){
                    comb = x[j:(j+k-1),'class']
                    start_vec = x[j:(j+k-1),'start']
                    if (order(comb[c(1,k)])[1]==2){
                        comb = comb[k:1]
                    }
                    comb_str = paste(comb,collapse='_')
                    if (k==2){
                        s1 = x[j,'strand']
                        s2 = x[j+1, 'strand']
                        if (s1 == s2){
                            comb_frame$ori[j] = 'same_strand'
                        } else if (s1=='-'){
                            comb_frame$ori[j] = 'outwards'
                        } else {
                            comb_frame$ori[j] = 'inwards'
                        }
                    }
                    comb_vec[j] = comb_str
                    dist_vec[j] = start_vec[k] - start_vec[1]
                    comb_frame[j,1:k] = comb
                    name_frame[j,] = x[j:(j+k-1),'name']
                    j = j + 1
                }
                return(data.frame(combination=comb_vec, dist_vec, comb_frame, name_frame))
            }
        })
    kmer_table = do.call(rbind.data.frame, kmer_list)
    return(kmer_table)
}
real_kmer = kmer_in_lad(o_table)

ggplot(real_kmer[which(real_kmer$class1!='boundary'&real_kmer$class2!='boundary'),],
       aes(x=combination, y=log10(dist_vec), fill=class1, color=class2)) + 
    geom_violin(alpha=0.5, size=2) +
    geom_point(aes(shape=ori), position=position_jitter(width=0.5), alpha=0.8) +
    ylab('log10(bp distance between TSS)') +
    scale_color_manual(values=COL_class) +
    scale_fill_manual(values=COL_class) +
    scale_shape_manual(values=c(same_strand=16, outwards=17, inwards=18)) 


ggplot(real_kmer[which(real_kmer$class1!='boundary'&real_kmer$class2!='boundary'),],
       aes(x=ori, fill=class1, color=class2)) + 
    geom_bar(alpha=0.5, size=2) +
    facet_wrap(~combination,nrow=1) +
    scale_color_manual(values=COL_class) +
    scale_fill_manual(values=COL_class)
```



```{r}
load('cl20170426_transcript_correlations_df.RData')
o = findOverlaps(lad_gr[lad_gr$name=='LAD'], tss_gr)

o_table = data.frame(o)
o_table$class = as.character(tss_gr[o_table$subjectHits]$class)
o_table$strand = as.vector(strand(tss_gr))[o_table$subjectHits]
o_table$start = as.vector(start(tss_gr))[o_table$subjectHits]
o_table$name = as.character(tss_gr[o_table$subjectHits]$name)
rownames(o_table) = gsub('[.][0-9]+','',as.character(tss_gr[o_table$subjectHits]$name))

test = mclapply(rownames(o_table),function(x,o_table, cor_df){
        other_table = o_table[rownames(o_table)!=x & rownames(o_table)%in%rownames(cor_df),]
        data.frame(same_lad = other_table$queryHits == o_table$queryHits[x],
                   correlation = cor_df[x,rownames(cor_df)])
    }, o_table=o_table, cor_df=cor_df, mc.cores=3)
```
